packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-choropleth-map/MapDataProvider.tsx

```diff
@@ -0,0 +1,19 @@
+import React, { useState, useEffect } from 'react';
+import generateMapData from './generateMapData';
+
+export default function MapDataProvider({
+  map,
+  children: Children,
+}: {
+  map: string;
+  children: React.ComponentType<{ data: Record<string, unknown>[] }>;
+}) {
+  const [data, setData] = useState<Record<string, unknown>[]>([]);
+  useEffect(() => {
+    generateMapData(map).then(mapData => {
+      setData(mapData);
+    });
+  }, [map]);
+
+  return <Children data={data} />;
+}
```

Comment:
Can this be a [custom hook](https://reactjs.org/docs/hooks-custom.html) instead? Just saw someone commenting on [React providers](https://twitter.com/DavidWells/status/1264742922190127104) earlier today...

```js
import { useState, useEffect } from 'react';

function useMap(map) {
  const [mapData, setMapData] = useState(null);

  useEffect(() => {
    generateMapData(map).then(mapData => {
      setData(mapData);
    });
  });

  return mapData;
}
```

```js
function WorldMap() {
  const mapData = useMap(select(
    'Map',
    maps.map(m => m.key),
    'world',
    'map',
  ));
  return mapData ? <SuperChart formData={{ map }} /> : null;
}
```

Reply:
Looks good. Will give this a try.